include/ginkgo/core/base/scoped_device_id.hpp

```diff
@@ -0,0 +1,227 @@
+ * sure that the device code is run on the correct device within one scoped
+ * block, when run with multiple devices. Depending on the executor it will
+ * record the current device id and set the device id to the one being passed
+ * in. After the scope has been exited, the destructor sets the device_id back
+ * to the one before entering the scope. The OmpExecutor and DpcppExecutor don't
+ * require setting an device id, so in those cases, the class is a no-op.
+ *
+ * The device id scope has to be constructed from a executor with concrete type
+ * (not plain Executor) and a device id. Only the type of the executor object is
+ * relevant, so the pointer will not be accessed, and may even be a nullptr.
+ * From the executor type the correct derived class of
+ * detail::generic_scoped_device_id is picked. The following illustrates the
+ * usage of this class:
+ * ```
+ * {
+ *   scoped_device_id g{static_cast<CudaExecutor>(nullptr), 1};
+ *   // now the device id is set to 1
+ * }
+ * // now the device id is reverted again
+ * ```
+ */
+class scoped_device_id {
+public:
+    /**
+     * Create a scoped device id from an OmpExecutor.
+     *
+     * This will pick the noop_scoped_device_id.
+     *
+     * @param exec  Not used.
+     * @param device_id  Not used.
+     */
+    scoped_device_id(const OmpExecutor* exec, int device_id)
+        : scope_(std::make_unique<detail::noop_scoped_device_id>())
+    {}
+
+    /**
+     * Create a scoped device id from an CudaExecutor.
+     *
+     * This will pick the cuda_scoped_device_id.
+     *
+     * @param exec  Not used.
+     * @param device_id  The device id to use within the scope.
+     */
+    scoped_device_id(const CudaExecutor* exec, int device_id)
+        : scope_(std::make_unique<detail::cuda_scoped_device_id>(device_id))
+    {}
+
+    /**
+     * Create a scoped device id from an HipExecutor.
+     *
+     * This will pick the hip_scoped_device_id.
+     *
+     * @param exec  Not used.
+     * @param device_id  The device id to use within the scope.
+     */
+    scoped_device_id(const HipExecutor* exec, int device_id)
+        : scope_(std::make_unique<detail::hip_scoped_device_id>(device_id))
+    {}
+
+    /**
+     * Create a scoped device id from an OmpExecutor.
+     *
+     * This will pick the noop_scoped_device_id.
+     *
+     * @param exec  Not used.
+     * @param device_id  Not used.
+     */
+    scoped_device_id(const DpcppExecutor* exec, int device_id)
+        : scope_(std::make_unique<detail::noop_scoped_device_id>())
+    {}
+
+    scoped_device_id() = default;
+
+    // Prohibit copy construction.
+    scoped_device_id(const scoped_device_id&) = delete;
+
+    // Allow move construction.
+    scoped_device_id(scoped_device_id&&) = default;
+
+    // Prohibit copy assignment.
+    scoped_device_id& operator=(const scoped_device_id&) = delete;
+
+    // Allow move construction.
+    scoped_device_id& operator=(scoped_device_id&&) = default;
```

Comment:
does a scope guard need to be movable?

Reply:
I thought I need it, because it is returned from a function and we can't rely on copy elision in that case. But I might be wrong. I will try again to use only copy elision.

Reply:
ah right, this is only guaranteed in C++17, so we probably need it for some older compilers